// newapi/src/SiteAtomic/Objects/Site/Site.ts
/**
 * Class Site
*/

import SerialisableObject from "../../../Framework/SerialisableObject";

export default class Site extends SerialisableObject {

    /**
     * The primary key for this site
     *
     * @var number 
     */
    private _siteId?:number;

    /**
     * The account id which this site belongs to.
     *
     * @var number 
     */
    private _accountId?:number;

    /**
     * A simple lowercase alphanumerical string which uniquely
     * identifies this site and also acts as the subdomain for
     * the control domain for this site.
     *
     * @var string 
     */
    private _siteKey?:string;

    /**
     * Site title (descriptive)
     *
     * @var string 
     */
    private _title?:string;



    /**
     * Constructor
     *
     * @param accountId
     * @param siteKey
     * @param title
     */
    constructor(accountId?:number, siteKey?:string, title?:string){

        super();

        this._accountId = accountId;
        this._siteKey = siteKey;
        this._title = title;
        
    }

    /**
     * Get the siteId
     *
     * @return number
     */
    public get siteId():number | undefined{

        return Number(this._siteId);
        
    }


    /**
     * Get the accountId
     *
     * @return number
     */
    public get accountId():number | undefined{

        return Number(this._accountId);
        
    }


    /**
     * Set the accountId
     *
     * @param number accountId
     * @return Site
     */
    public set AccountId(accountId:number | undefined){
        this._accountId = accountId;
    }

    /**
     * Get the siteKey
     *
     * @return string
     */
    public get siteKey():string | undefined{

        
        return this._siteKey;
    }


    /**
     * Set the siteKey
     *
     * @param string siteKey
     * @return Site
     */
    public set SiteKey(siteKey:string | undefined){
        this._siteKey = siteKey;
    }

    /**
     * Get the title
     *
     * @return string
     */
    public get title():string | undefined{

        
        return this._title;
    }


    /**
     * Set the title
     *
     * @param string title
     * @return Site
     */
    public set Title(title:string | undefined){
        this._title = title;
    }


}
